### PR TITLE
Add histograms for checkpoint and state transfer duration

### PR DESCRIPTION
**Pull request by yontyon**
Add multiple histograms:
1. To measure the time that passes between a creation of a checkpoint to the time it became stable
2. To measure the time that state transfer takes from starting to its end

### CODE REVIEW

File: bftengine/src/bftengine/ReplicaImp.hpp [hunk 1]
```diff
@@ -463,7 +463,8 @@ class ReplicaImp : public InternalReplicaApi, public ReplicaForStateTransfer {
                                         {"onSeqNumIsStable", onSeqNumIsStable},
                                         {"onTransferringCompleteImp", onTransferringCompleteImp},
                                         {"consensus", consensus},
-                                        {"timeInActiveView", timeInActiveView}});
+                                        {"timeInActiveView", timeInActiveView},
+                                        {"checkpointFromCreationToStable", checkpoint_creation_to_stable}});
```
Comment (andrewjstone) [inline bftengine/src/bftengine/ReplicaImp.hpp:467]: It looks like you missed adding `timeInStateTransfer` here.

File: bftengine/src/bftengine/ReplicaImp.hpp [hunk 2]
```diff
@@ -490,16 +491,23 @@ class ReplicaImp : public InternalReplicaApi, public ReplicaForStateTransfer {
 
     std::shared_ptr<Recorder> timeInActiveView =
         std::make_shared<Recorder>(1, MAX_VALUE_SECONDS, 3, concord::diagnostics::Unit::SECONDS);
+
+    std::shared_ptr<Recorder> timeInStateTransfer =
```
Comment (andrewjstone) [inline bftengine/src/bftengine/ReplicaImp.hpp:495]: I'm not sure 5 minutes is a long enough duration here. Maybe make it a few hours for max? I wouldn't change the constant above though, as making the duration longer makes the histograms larger and for most histograms we don't need hours.

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 3]
```diff
@@ -3775,6 +3779,7 @@ void ReplicaImp::executeRequestsInPrePrepareMsg(concordUtils::SpanWrapper &paren
   if ((lastExecutedSeqNum + 1) % checkpointWindowSize == 0) {
     checkpointNum = (lastExecutedSeqNum + 1) / checkpointWindowSize;
     stateTransfer->createCheckpointOfCurrentState(checkpointNum);
+    checkpoint_times_.start(checkpointNum);
```
Comment (andrewjstone) [inline bftengine/src/bftengine/ReplicaImp.cpp:3782]: It appears that you are using checkpoint numbers here which are sequence numbers divided by window size. However, in the end method you are using sequence numbers directly. These will not match up.

**Review by andrewjstone — CHANGES_REQUESTED** [review]
Hi @yontyon thanks for the additions. Unfortunately there are a few errors here that will prevent this code from working as intended. After fixing I suggest you run a cluster in a docker environment, induce state transfer via stopping and starting a node, and then sanity check with `./concord-ctl perf get`.

File: bftengine/src/bftengine/ReplicaImp.hpp (hunk 2 above)
Comment (yontyon) [reply bftengine/src/bftengine/ReplicaImp.hpp:495]: MAX_VALUE_SECONDS is set to 5 years, Am I missing something?

File: bftengine/src/bftengine/ReplicaImp.hpp (hunk 1 above)
Comment (yontyon) [reply bftengine/src/bftengine/ReplicaImp.hpp:468]: fixed

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 3 above)
Comment (yontyon) [reply bftengine/src/bftengine/ReplicaImp.cpp]: fixed

File: bftengine/src/bftengine/ReplicaImp.hpp (hunk 2 above)
Comment (andrewjstone) [reply bftengine/src/bftengine/ReplicaImp.hpp:505]: Whoops. I misread. Sorry.

**Review by andrewjstone — APPROVED** [review]
Looks good @yontyon. Thanks for the fixes and the new histograms! 

I only had one minor comment about using declarations. We aren't supposed to have them in header files unless encapsulated in some manner so they don't infect users.

File: bftengine/src/bftengine/CollectorOfThresholdSignatures.hpp [hunk 4]
```diff
@@ -27,6 +27,9 @@
 #include "Logger.hpp"
 #include "kvstream.h"
 #include "demangle.hpp"
+#include "diagnostics.h"
+
+using Recorder = concord::diagnostics::Recorder;
```
[the remark this answers is not in the document]
Comment (yontyon) [reply bftengine/src/bftengine/CollectorOfThresholdSignatures.hpp]: done

**Comment by yontyon** [comment]
After some sanity checks, it looks like the digital signatures operation histograms are not accurate.
Thus, we reset the PR to have only checkpoint and state transfer metrics.
The DS histograms will be added in a different PR

**Pull request merged**